### ISSUE TITLE
Consider making some reformatting configurable

### ISSUE THREAD

**Issue by TomasHubelbauer** [slang800/tidy-markdown#6]
Since this library is a dependency of [`atom-tidy-markdown`](https://github.com/slang800/atom-tidy-markdown), but is the actual implementation of the linting, I'd like to introduce an idea of making some reformattings configurable and having those options be editable in the plugin settings page.

Namely I'd be interested in seeing empty lines between headings allowable to stay intact / allowable to collapse to a single empty line, either way, not removed at all.

Please let me know what you think about the suggestion.

**Comment by slang800** [comment]
If there's a good reason for having empty lines between headings, then I can change the style-guide that this is built off of to reflect that reason.

However, I want to avoid adding options just for people's preferences. The idea is to output a canonical representation of a markdown document... A single "Right Way" to do it... So having extra configuration would defeat that purpose.

**Comment by TomasHubelbauer** [comment]
There is not, having empty lines between headings is purely a preference. That was just an example of what I'm missing the most, though, the general idea was to provide some common settings for adjusting most common preferences and seeing how this goes against the idea of a single set of formatting rules, which I have not been aware of being by design, I'm closing this issue. Thanks!

**Comment by benbalter** [comment]
> If there's a good reason for having empty lines between headings, then I can change the style-guide that this is built off of to reflect that reason.

To me, Markdown is fundamentally "how you'd format something on a typewriter (or in notepad) when you have only plain text and no formatting options". In that world (and even in microsoft word when you do) headings are generally followed by an empty line to separate them from the surrounding text and for increased visibility. While not deterministic, take a look at [this excerpt from typography for lawyers](http://typographyforlawyers.com/hard-line-breaks.html) which itself has vertical space after the paragraph when rendered (which we'd want markdown to represent as closely as possible).

**Comment by TomasHubelbauer** [comment]
That book looks very interesting, I'm going to buy it. Thanks for commenting on this.

**Comment by benbalter** [comment]
@TomasHubelbauer highly recommend it (I have it in both print and Kindle). The author, @mbutterick is even a GitHub user!

**Comment by mbutterick** [comment]
De-lawyerized version also available at [Practical Typography](http://practicaltypography.com)

**Comment by slang800** [comment]
@benbalter: That's a good point - I'll look into this further.

**Comment by Glavin001** [comment]
[Aom Beautify is switching from Pandoc to tidy-markdown](https://github.com/Glavin001/atom-beautify/issues/292), and it would be great to have more customizable options for users. Thanks for the great work so far!

**Comment by josh-egan-ps** [comment]
I'm using this with Atom Beautify, and I would love to see some configurable options! I don't like the feature that converts my h3 into an h2. In general, I think a beautifier should format and 'beautify' my code, not change it. I can understand some people wanting this feature, but I would like to turn it off. 

Here's an example: 
```
# Header

### Sub Header
```

Get's turned into:
```
# Header
## Sub Header
```

So I lose my whitespace and my h3 is converted to an h2. As mentioned above, I would also like to be able to customize whitespace. I would recommend having a configuration value of "Maximum New Lines to Preserve" or something like that (see Javascript - Max preserve newlines as an example in atom-beautify.

**Comment by josh-egan** [comment]
I'm using this with Atom Beautify, and I would love to see some configurable options! I don't like the feature that converts my h3 into an h2. In general, I think a beautifier should format and 'beautify' my code, not change it. I can understand some people wanting this feature, but I would like to turn it off. 

Here's an example: 
```
# Header

### Sub Header
```

Get's turned into:
```
# Header
## Sub Header
```

So I lose my whitespace and my h3 is converted to an h2. As mentioned above, I would also like to be able to customize whitespace. I would recommend having a configuration value of "Maximum New Lines to Preserve" or something like that. See `Javascript - Max preserve newlines` as an example in atom-beautify.

**Comment by slang800** [comment]
@josh-egan: out of curiosity, why do you jump from h1 to h3?

**Comment by slang800** [comment]
Today I did an informal survey of markdown docs, and it seems like space after headers and before lists is almost universal:

name                  | \n after header | \n before lists | ul char    | hardwrap
--------------------- | --------------- | --------------- | ---------- | --------
Node.js Docs          | true            | true            | *          | true
Ruby Docs (not MD)    | true            | true            | *          | true
CommonMark Spec       | true            | true            | -          | true
Python Docs  (not MD) | true            | true            | *          | true
When.js Docs          | mixed           | mixed           | *          | false
Bootstrap Readme      | true            | true            | -          | false
Jade Readme           | true            | true            | -          | false
Stylus Docs           | true            | true            | -          | false
Atom Editor Docs      | true            | true            | mixed      | false
Piwik Docs            | true            | true            | -          | false
Discourse Docs        | true            | true            | -          | false
Docker Docs           | true (mostly)   | true            | -          | true
Puppet Docs           | true            | true            | *          | true
Hubot Docs            | true            | true            | *          | false
PredictionIO Manual   | true            | true            | - (mostly) | false

Because of that (along with @benbalter's reasoning) I'mma change the styleguide & this tool to put whitespace in those places.

**Comment by josh-egan** [comment]
@slang800 Just because I prefer the h3 styling on Github better. The h2 tag is followed by a line on github, and when there are frequent headers I think the lines introduce clutter. For example, I use h3 tags intstead of h2 tags in this document because of this visual preference: https://github.com/josh-egan/Configurations/blob/master/Git/configurations.md

**Comment by slang800** [comment]
@josh-egan: Ah, that annoyed me too... I actually made a chrome & firefox plugin called [gh-fixed](https://github.com/slang800/gh-fixed) to deal with that (since I didn't want to alter the structure of my documents just to change the style). Or, if you don't want the whole plugin, you can just get the H2 change by applying this style:

```css
.markdown-body h2 {
  border-bottom: none !important;
  padding-bottom: 0 !important;
}
```

...via [Stylish](https://userstyles.org/styles/114864/remove-lines-under-h2s-on-github) or something like that.

**Comment by josh-egan** [comment]
That's an interesting idea. I might consider that for myself. The challenge is that I can't expect other people to do something special for formatting, and I want it to look decent when anyone else sees it too. What do you think about making it configurable?

**Comment by rugk** [comment]
I would also like to see an option to use another character for listings. E.g. `*`instead of `-` which is forced currently.

**Comment by jbnv** [comment]
> I want to avoid adding options just for people's preferences. The idea is to output a canonical representation of a markdown document... A single "Right Way" to do it... So having extra configuration would defeat that purpose.

This is an arrogant statement that makes me not want to use this code or support this project. Maybe fork it and modify it to my liking, but that's it. The "canonical" changes that you make to my markdown are counterproductive. I have my markdown formatted the way that I wanted for a reason.

**Comment by slang800** [comment]
> This is an arrogant statement that makes me not want to use this code or support this project. Maybe fork it and modify it to my liking, but that's it.

It's GPLv3, so feel free.

> The "canonical" changes that you make to my markdown are counterproductive.

Are you arguing against canonicalization as an idea, or just the set of rules that tidy-markdown uses?

> I have my markdown formatted the way that I wanted for a reason.

If you can elaborate on & justify the reason for whatever formatting you apply, we can get it written up and merged into the [styleguide](https://github.com/slang800/tidy-markdown). I don't think that markdown formatting habits are the result of irreconcilable differences between people or the projects that markdown is used within. We should be able to define a single "Right Way" that is objectively superior for the vast majority of users.

**Comment by jbnv** [comment]
> Are you arguing against canonicalization as an idea, or just the set of rules that tidy-markdown uses?

Be canonical if you want, if that fits your purposes. I use Markdown because it's about as lightweight as markup can get. I don't care about canon. I'm going to run the content through HTML converters or other tools; nobody but me is going to see or use my content in Markdown.

> If you can elaborate on & justify the reason for whatever formatting you apply ... We should be able to define a single "Right Way" that is objectively superior for the vast majority of users.

Why does it matter to you how I format my content? Why is it more important that I conform to some canon no matter whether or how much the canon interferes with my workflow?

**Comment by slang800** [comment]
First off, I should point out that if you don't agree with canonicalization of data, or don't think that it should be applied to markdown, you're free to not use tidy-markdown. It's not built-in to Atom, or even installed by default... I really just made it for people who admire the automation of formatting that's present in tools like [gofmt](https://golang.org/cmd/gofmt/) and want to apply that to markdown too. However, I'll present an argument as to why it's a good idea in general, because I love a good debate.

Documents are read more often than they are written and by more people than just the authors, much like code. Thus, it stands to reason that we should optimize documents for readability, to make them easier to proofread and update. All of the rules in the [styleguide](https://github.com/slang800/tidy-markdown) have been chosen with readability in mind, so using these rules should produce a highly readable document.

But there's an additional aspect of using a styleguide that helps readability too: consistency. By formatting all documents in the same way, and each type of element within that document in the same way, it becomes easier to parse. With consistent style, you can give less thought to the individual characters and syntax of the document, and locate parts of the document simply by the layout. This is analogous to how syntax highlighting makes it easier to read and find parts of documents just by color.

In addition, canonicalization removes redundant syntax. For example, you don't need to remember that `# header`, `header\n======`, and `# header #` are all the exact same thing. Instead, you only encounter the common (atx-style) `# header`.

And when you only use a single syntax for each element, it becomes easier to edit. You can use ad-hoc regexes like `/^-\s/` to find the beginnings of all the lists in a document (or several), or place cursors at the beginning of all the headers in a document with `/^#+\s/`.

Finally, having a styleguide reduces [decision fatigue](https://en.wikipedia.org/wiki/Decision_fatigue) by removing all stylistic decisions.

**Comment by jbnv** [comment]
There's nothing to debate. This plug-in makes changes to my content that are counterproductive. We asked for options to let us work around those things. You refuse to implement those options. Therefore I choose to not use this plug-in.

If you are interested in humoring me, explain how removing the whitespace between a #-header and the paragraph beneath it makes the text more readable. Ditto why you have the moral right to remove whitespace between bullet items that the user explicitly put in there for whatever reason. These are the two main problems that I have with this plug-in.

**Comment by rugk** [comment]
> If you are interested in humoring me, explain how removing the whitespace between a #-header and the paragraph beneath it makes the text more readable. Ditto why you have the moral right to remove whitespace between bullet items that the user explicitly put in there for whatever reason. These are the two main problems that I have with this plug-in.

This seems to be fixed in https://github.com/slang800/tidy-markdown/issues/25.

**Comment by jbnv** [comment]
I have version 1.0.0. It does both of those things.

**Comment by rugk** [comment]
Yeah, because https://github.com/slang800/tidy-markdown/pull/25 was not merged yet... :smile:

**Comment by rugk** [comment]
@jbnv BTW there is no v1.0.0. yet [according to the releases page](https://github.com/slang800/tidy-markdown/releases).
@slang800 Will you release a new version with the fix of #25?

**Comment by josh-egan-ps** [comment]
@slang800 I wanted to revisit the conversation about not enforcing headings to go from #, to ##, to ### in a document. I came across this use case today:

I have a blog where the h1 heading is generated by a top level template from json data, and the content of the article comes from a markdown file. In the markdown file, I want to start the headings at h2, because the h1 will be generated by the template at the top of the page. However, tidy-markdown enforces that a h1 be present within the document, so it is causing me problems. For the time being, I have an h1 at the top just to allow me to beautify the document as I go, and I'm going to need to remember to remove it before the post goes live. I've also had to disable automatic beautifying so that I can save the document after removing the h1 and not have it get messed up.

Thoughts?
